### PR TITLE
Story/rhoaieng 2224 convert api list pipelines pipeline versions

### PR DESCRIPTION
**Pull request by Gkrumbach07**
closes: https://issues.redhat.com/browse/RHOAIENG-2224

## Description

sorry this is a big one bc list pipelines hits a lot of files. no big changes tho
- also testing this comment (version upload): https://github.com/opendatahub-io/odh-dashboard/pull/2417#pullrequestreview-1854882319
- delete pipeline now has new logic -> must delete all pipeline versions before you can delete the pipeline. This was bc of a change in v2 api. We disable the kebab delete button now when versions > 0.
- updated types

## How Has This Been Tested?

1. create a v2 dspa
2. import a pipeline
3. import a version for that
4. notice delete kebab is disabled for the pipeline
5. delete both versions
6. delete pipeline from now enabled kebab

## Test Impact

covered under versions tests

## Request review criteria:

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit tests & storybook for related changes)

If you have UI changes: 

- [x] Included any necessary screenshots or gifs if it was a UI change.
- [x] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

### CODE REVIEW

**Review by jpuzz0 — COMMENTED** [review]
This looks good. 

I do have a concern with the change to the delete logic as described in the PR summary. I think at least a tooltip might be needed for the disabled action so that users know they need to delete their versions before being able to delete the pipeline they're contained in. Maybe @yannnz could chime in.

File: frontend/src/concepts/pipelines/content/DeletePipelineModalExpandableSection.tsx [hunk 1]
```diff
@@ -58,7 +58,7 @@ const DeletePipelineModalExpandableSection: React.FC<DeletePipelineModalExpandab
           }
 
           return (
-            <ListItem key={resource.id} icon={icon}>
+            <ListItem key={resource.display_name} icon={icon}>
```
[the remark this answers is not in the document]
Comment (Gkrumbach07) [reply frontend/src/concepts/pipelines/content/DeletePipelineModalExpandableSection.tsx:61]: ah but this is from a resource that is not necessarily a pipeline and kf doesn't have a common unique id anymore

File: frontend/src/concepts/pipelines/apiHooks/useAllPipelineVersions.ts [hunk 2]
```diff
@@ -13,12 +13,14 @@ import { useDeepCompareMemoize } from '~/utilities/useDeepCompareMemoize';
 export const useAllPipelineVersions = (
   options: PipelineOptions = {},
   refreshRate = 0,
-): FetchState<PipelineListPaged<PipelineVersionKF>> => {
+): FetchState<PipelineListPaged<PipelineVersionKFv2>> => {
   const { api } = usePipelinesAPI();
   const [{ items: pipelines }] = usePipelines();
-  const pipelineIds = useDeepCompareMemoize(pipelines?.map((pipeline) => pipeline.id) || []);
+  const pipelineIds = useDeepCompareMemoize(
+    pipelines?.map((pipeline) => pipeline.pipeline_id) || [],
```
[the remark this answers is not in the document]
Comment (Gkrumbach07) [reply frontend/src/concepts/pipelines/apiHooks/useAllPipelineVersions.ts:20]: thats the linter. cannot remove

File: frontend/src/concepts/pipelines/content/DeletePipelineModalExpandableSection.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (Gkrumbach07) [reply frontend/src/concepts/pipelines/content/DeletePipelineModalExpandableSection.tsx]: i added the pos as part of it now. not the best but i dont think this list will do much rer-rendering bc it is the delete modal

**Review by jpuzz0 — APPROVED** [review]
/lgtm

File: frontend/src/concepts/pipelines/apiHooks/usePipelineQuery.ts [hunk 3]
```diff
@@ -8,7 +12,8 @@ import { K8sAPIOptions } from '~/k8sTypes';
 export type PipelineKFCallCommonWithItems<T extends PipelineCoreResourceKF> =
   PipelineKFCallCommon<unknown> & { items?: T[] };
 
-const usePipelineQuery = <T extends PipelineCoreResourceKF>(
+// TODO: remove this OR when we remove all the old pipeline types (PipelineCoreResourceKF)
+const usePipelineQuery = <T extends PipelineCoreResourceKFv2 | PipelineCoreResourceKF>(
```
Comment (andrewballantyne) [inline frontend/src/concepts/pipelines/apiHooks/usePipelineQuery.ts:16]: Curious -- what still uses this that is on v1? The singular get?

File: frontend/src/concepts/pipelines/content/DeletePipelineModalExpandableSection.tsx [hunk 4]
```diff
@@ -58,7 +58,7 @@ const DeletePipelineModalExpandableSection: React.FC<DeletePipelineModalExpandab
           }
 
           return (
-            <ListItem key={resource.id} icon={icon}>
+            <ListItem key={resource.display_name + i} icon={icon}>
```
Comment (andrewballantyne) [inline frontend/src/concepts/pipelines/content/DeletePipelineModalExpandableSection.tsx:61]: Interesting -- that's inconvenient... I wonder if we can make a utility that reads properties and returns an id -- I don't like the idea of index because it leads to react slowness. 

if you use an integer in each key, React de-optimizes and rerenders the whole list each time -- instead of just the element that changed.

File: frontend/src/concepts/pipelines/content/tables/pipeline/PipelinesTableRow.tsx [hunk 5]
```diff
@@ -110,28 +111,25 @@ const PipelinesTableRow: React.FC<PipelinesTableRowProps> = ({
         </Tr>
         {isExpanded && (
           <PipelinesTableExpandedRow
-            // Upload a new version will update the pipeline default version
-            // Which could trigger a re-render of the versions table
-            key={pipeline.default_version?.id}
+            // Upload a new version will update the totalSize (version length)
+            // Which will trigger a re-render of the versions table
+            key={`${pipeline.pipeline_id}-expanded-${totalSize}`}
```
Comment (andrewballantyne) [inline frontend/src/concepts/pipelines/content/tables/pipeline/PipelinesTableRow.tsx:116]: Wait, why does this need to happen? I get this logic is coming from Versions -- but what? @DaoDaoNoCode why did you need to do this?

File: frontend/src/concepts/pipelines/content/tables/pipelineRun/PipelineRunTableRow.tsx [hunk 6]
```diff
@@ -47,12 +47,13 @@ const PipelineRunTableRow: React.FC<PipelineRunTableRowProps> = ({
         <CoreResourceExperiment resource={run} />
       </Td>
       <Td modifier="truncate" dataLabel="Pipeline">
-        <CoreResourcePipelineVersion
+        {/* TODO: bring back with pipeline runs */}
```
Comment (andrewballantyne) [inline frontend/src/concepts/pipelines/content/tables/pipelineRun/PipelineRunTableRow.tsx:50]: Ticket please

File: frontend/src/concepts/pipelines/content/tables/pipelineRun/PipelineRunTableToolbar.tsx [hunk 7]
```diff
@@ -56,13 +56,15 @@ const PipelineRunTableToolbar: React.FC<PipelineRunJobTableToolbarProps> = ({
...
           />
         ),
-        [FilterOptions.PIPELINE_VERSION]: ({ onChange, label }) => (
-          <PipelineVersionSelect
-            versions={versions}
-            selection={label}
-            onSelect={(version) => onChange(version.id, version.name)}
-          />
-        ),
+        // TODO: Uncomment this when adding runs for v2
```
Comment (andrewballantyne) [inline frontend/src/concepts/pipelines/content/tables/pipelineRun/PipelineRunTableToolbar.tsx:59]: ticket please

File: frontend/src/concepts/pipelines/content/tables/pipelineRunJob/PipelineRunJobTableRow.tsx [hunk 8]
```diff
@@ -48,12 +48,13 @@ const PipelineRunJobTableRow: React.FC<PipelineRunJobTableRowProps> = ({
         <CoreResourceExperiment resource={job} />
       </Td>
       <Td modifier="truncate" dataLabel="Pipeline">
-        <CoreResourcePipelineVersion
+        {/* TODO: bring back with pipeline runs */}
```
Comment (andrewballantyne) [inline frontend/src/concepts/pipelines/content/tables/pipelineRunJob/PipelineRunJobTableRow.tsx:51]: ticket please

File: frontend/src/concepts/pipelines/content/tables/pipelineRunJob/PipelineRunJobTableToolbar.tsx [hunk 9]
```diff
@@ -51,13 +51,15 @@ const PipelineRunJobTableToolbar: React.FC<PipelineRunJobTableToolbarProps> = ({
...
         ),
-        [FilterOptions.PIPELINE_VERSION]: ({ onChange, label }) => (
-          <PipelineVersionSelect
-            versions={versions}
-            selection={label}
-            onSelect={(version) => onChange(version.id, version.name)}
-          />
-        ),
+        [FilterOptions.PIPELINE_VERSION]: () => <></>,
+        // TODO: Uncomment this when adding runs for v2
```
Comment (andrewballantyne) [inline frontend/src/concepts/pipelines/content/tables/pipelineRunJob/PipelineRunJobTableToolbar.tsx:55]: ticket please -- not sure why you stopped doing this lol

File: frontend/src/concepts/pipelines/content/useDeleteStatuses.tsx [hunk 10]
```diff
@@ -39,9 +40,19 @@ const useDeleteStatuses = ({
...
-            <b>{resource.name}</b>: {state.message}
+          <p
+            key={
+              (resource as PipelineCoreResourceKFv2).display_name ??
+              (resource as PipelineCoreResourceKF).name
+            }
+          >
+            <b>
+              {(resource as PipelineCoreResourceKFv2).display_name ??
+                (resource as PipelineCoreResourceKF).name}
```
Comment (andrewballantyne) [inline frontend/src/concepts/pipelines/content/useDeleteStatuses.tsx:53]: Not that it overly matters -- but next time make a variable lol, they don't cost much.

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
Minor updates, testing now.

**Comment by andrewballantyne** [comment]
> I do have a concern with the change to the delete logic as described in the PR summary. I think at least a tooltip might be needed for the disabled action so that users know they need to delete their versions before being able to delete the pipeline they're contained in. Maybe @yannnz could chime in.

Great point @jpuzz0 -- please add the tooltip in this PR.

File: frontend/src/concepts/pipelines/content/useDeleteStatuses.tsx (hunk 10 above)
Comment (Gkrumbach07) [reply frontend/src/concepts/pipelines/content/useDeleteStatuses.tsx:53]: oh this is temp code tho. just wanted to make minimal impact

File: frontend/src/concepts/pipelines/content/tables/pipeline/PipelinesTableRow.tsx (hunk 5 above)
Comment (andrewballantyne) [reply frontend/src/concepts/pipelines/content/tables/pipeline/PipelinesTableRow.tsx:116]: That is not making any sense to me lol -- why wouldn't your data just update inline? This is React...

File: frontend/src/concepts/pipelines/content/utils.tsx [hunk 11]
```diff
@@ -87,6 +91,9 @@ export const getPipelineIdByPipelineVersion = (
   )?.key.id;
 
 export const getPipelineAndVersionDeleteString = (
-  resources: PipelineCoreResourceKF[],
+  resources: PipelineCoreResourceKFv2[],
   type: 'pipeline' | 'version',
 ): string => `${resources.length} ${type}${resources.length !== 1 ? 's' : ''}`;
+
+export const getPipelineResourceUniqueID = (resource: PipelineCoreResourceKFv2): string =>
+  resource.display_name + resource.created_at;
```
Comment (andrewballantyne) [inline frontend/src/concepts/pipelines/content/utils.tsx:99]: Why not make this use the id of the resource? Although I don't mind this solution -- you could have done this at the top level -- why did you use index? lol

**Review by andrewballantyne — APPROVED** [review]
Looks good... we need to deal with the checkbox but it can be a different ticket (will log it after I approve here)

EDIT: https://issues.redhat.com/browse/RHOAIENG-2574
